gateway-service-impl/src/main/java/org/hypertrace/gateway/service/baseline/BaselineServiceImpl.java

```diff
@@ -0,0 +1,383 @@
+                  metricValues.stream().mapToDouble(Double::doubleValue).toArray();
+              Baseline baseline = BaselineCalculator.getBaseline(metricValueArray);
+              baselineMap.put(key, baseline);
+            });
+        baselineEntityBuilder.putAllBaselineAggregateMetric(baselineMap);
+        baselineEntityMap.put(baselineEntity.getId(), baselineEntityBuilder.build());
+      }
+    }
+    return baselineEntityMap;
+  }
+
+  private List<TimeAggregation> getTimeAggregationsForTimeSeriesExpr(
+      BaselineEntitiesRequest originalRequest) {
+    List<BaselineTimeAggregation> timeSeriesList =
+        originalRequest.getBaselineMetricSeriesRequestList();
+    List<TimeAggregation> timeAggregations = new ArrayList<>();
+    for (BaselineTimeAggregation timeAggregation : timeSeriesList) {
+      Expression aggregation =
+          Expression.newBuilder().setFunction(timeAggregation.getAggregation()).build();
+      TimeAggregation agg =
+          TimeAggregation.newBuilder()
+              .setAggregation(aggregation)
+              .setPeriod(timeAggregation.getPeriod())
+              .build();
+      timeAggregations.add(agg);
+    }
+    return timeAggregations;
+  }
+
+  private List<TimeAggregation> getTimeAggregationsForAggregateExpr(
+      BaselineEntitiesRequest originalRequest) {
+    List<FunctionExpression> aggregateList = originalRequest.getBaselineAggregateRequestList();
+    List<TimeAggregation> timeAggregationList = new ArrayList<>();
+    for (FunctionExpression function : aggregateList) {
+      TimeAggregation timeAggregation =
+          getAggregationFunction(
+              function, originalRequest.getStartTimeMillis(), originalRequest.getEndTimeMillis());
+      timeAggregationList.add(timeAggregation);
+    }
+    return timeAggregationList;
+  }
+
+  private TimeAggregation getAggregationFunction(
+      FunctionExpression function, long startTimeMillis, long endTimeMillis) {
+    Period period = getPeriod(startTimeMillis, endTimeMillis);
+    TimeAggregation timeAggregation =
+        TimeAggregation.newBuilder()
+            .setPeriod(period)
+            .setAggregation(Expression.newBuilder().setFunction(function).build())
+            .build();
+    return timeAggregation;
+  }
+
+  private Period getPeriod(long startTimeInMillis, long endTimeMillis) {
+    Period period;
+    if ((endTimeMillis - startTimeInMillis) < DAY_IN_MILLIS) {
+      long timePeriod = (endTimeMillis - startTimeInMillis) / (1000 * 60);
+      period = Period.newBuilder().setUnit("MINUTES").setValue((int) timePeriod).build();
+    } else {
+      long timePeriod = (endTimeMillis - startTimeInMillis) / (1000 * 60 * 60 * 24);
+      period = Period.newBuilder().setUnit("HOURS").setValue((int) timePeriod).build();
+    }
+    return period;
+  }
+
+  private Period getTimeSeriesPeriod(List<BaselineTimeAggregation> baselineTimeSeriesList) {
+    return baselineTimeSeriesList.get(0).getPeriod();
```

Comment:
Responded elsewhere on this - we need to handle the no data case without erroring.

Reply:
It is handled properly, if (originalRequest.getBaselineMetricSeriesRequestCount() > 0) in calling method. In that case empty response will be returned.

Reply:
@pavan-traceable what happens when query-service returns an empty response?

Reply:
ok - I traced this back and I believe the confusion to be a consequence of poor variable names - I assumed baselineTimeSeriesList was a list of timeSeries rather than a list of timeSeriesRequests. I think this could benefit from a rename, but not a blocking issue.

Reply:
updated variable name